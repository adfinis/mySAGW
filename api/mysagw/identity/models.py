from django.conf import settings
from django.contrib.postgres.fields import DateRangeField
from django.db import models
from django.db.models import Q
from django.utils import timezone
from django_countries.fields import CountryField
from localized_fields.fields import LocalizedCharField, LocalizedTextField
from phonenumber_field.modelfields import PhoneNumberField

from mysagw.models import HistoricalModel, TrackingModel, UniqueBooleanField, UUIDModel


class InterestCategory(UUIDModel, HistoricalModel):
    title = LocalizedCharField()
    description = models.CharField(max_length=255, blank=True, null=True)
    archived = models.BooleanField(default=False)
    public = models.BooleanField(default=False)


class Interest(UUIDModel, HistoricalModel):
    title = LocalizedCharField()
    description = models.CharField(max_length=255, blank=True, null=True)
    category = models.ForeignKey(
        InterestCategory, related_name="interests", on_delete=models.PROTECT
    )
    archived = models.BooleanField(default=False)


class MembershipRole(UUIDModel, HistoricalModel):
    title = LocalizedCharField()
    description = LocalizedTextField(blank=True, null=True, required=False)
    archived = models.BooleanField(default=False)


class Membership(UUIDModel, HistoricalModel):
    identity = models.ForeignKey(
        "Identity",
        on_delete=models.CASCADE,
        related_name="memberships",
    )
    organisation = models.ForeignKey(
        "Identity",
        on_delete=models.CASCADE,
        related_name="members",
    )
    role = models.ForeignKey(
        MembershipRole,
        related_name="memberships",
        on_delete=models.SET_NULL,
        null=True,
        blank=True,
    )
    authorized = models.BooleanField(default=False)
    time_slot = DateRangeField(null=True, blank=True)
    next_election = models.DateField(null=True, blank=True)
    comment = models.TextField(blank=True, null=True)
    inactive = models.BooleanField(default=False)


class Identity(UUIDModel, HistoricalModel, TrackingModel):
    SALUTATION_MR = "male"
    SALUTATION_MRS = "female"
    SALUTATION_NEUTRAL = "neutral"

    SALUTATION_LOCALIZED_MAP = {
        SALUTATION_MR: {"de": "Herr", "en": "Mr.", "fr": "Monsieur"},
        SALUTATION_MRS: {"de": "Frau", "en": "Mrs.", "fr": "Madame"},
        SALUTATION_NEUTRAL: {"de": "", "en": "", "fr": ""},
    }

    SALUTATION_CHOICES = (
        (SALUTATION_MR, SALUTATION_MR),
        (SALUTATION_MRS, SALUTATION_MRS),
        (SALUTATION_NEUTRAL, SALUTATION_NEUTRAL),
    )

    idp_id = models.CharField(max_length=255, unique=True, null=True, blank=False)
    email = models.EmailField(unique=True, null=True, blank=True)
    organisation_name = models.CharField(max_length=255, null=True, blank=True)
    first_name = models.CharField(max_length=255, null=True, blank=True)
    last_name = models.CharField(max_length=255, null=True, blank=True)
    salutation = models.CharField(
        choices=SALUTATION_CHOICES, default=SALUTATION_NEUTRAL, max_length=7
    )
    language = models.CharField(
        max_length=2, choices=settings.LANGUAGES, default=settings.LANGUAGE_CODE
    )
    interests = models.ManyToManyField(Interest, related_name="identities", blank=True)
    is_organisation = models.BooleanField(default=False)
    is_expert_association = models.BooleanField(default=False)
    is_advisory_board = models.BooleanField(default=False)
    comment = models.TextField(blank=True, null=True)

    def _get_memberships(self, only_authorized=False):
        memberships = Membership.objects.filter(
            Q(identity=self),
            Q(inactive=False),
            Q(time_slot__isnull=True) | Q(time_slot__contains=timezone.now()),
        )
        if only_authorized:
            memberships = memberships.filter(authorized=True)
        return self.__class__.objects.filter(
            pk__in=memberships.values_list("organisation", flat=True).distinct()
        )

    @property
    def authorized_for(self):
        return self._get_memberships(only_authorized=True)

    @property
    def member_of(self):
        return self._get_memberships(only_authorized=False)

    @property
    def localized_salutation(self):
        return self.SALUTATION_LOCALIZED_MAP[self.salutation][self.language]

    class Meta:
        ordering = ("last_name", "first_name", "email")


class Email(UUIDModel, HistoricalModel):
    identity = models.ForeignKey(
        Identity, related_name="additional_emails", on_delete=models.CASCADE
    )
    email = models.EmailField()
    description = LocalizedCharField(blank=True, null=True, required=False)

    class Meta:
        unique_together = [["identity", "email"]]
        ordering = ("email",)


class PhoneNumber(UUIDModel, HistoricalModel):
    identity = models.ForeignKey(
        Identity, related_name="phone_numbers", on_delete=models.CASCADE
    )
    phone = PhoneNumberField()
    description = LocalizedCharField(blank=True, null=True, required=False)
    default = UniqueBooleanField(default=False, together=["identity"])

    class Meta:
        unique_together = [["identity", "phone"]]
        ordering = ("-default", "phone")


class Address(UUIDModel, HistoricalModel):
    identity = models.ForeignKey(
        Identity, related_name="addresses", on_delete=models.CASCADE
    )
    address_addition_1 = models.CharField(max_length=255, null=True, blank=True)
    address_addition_2 = models.CharField(max_length=255, null=True, blank=True)
    address_addition_3 = models.CharField(max_length=255, null=True, blank=True)
    street_and_number = models.CharField(max_length=255)
    po_box = models.CharField(max_length=255, null=True, blank=True)
    postcode = models.CharField(max_length=255)
    town = models.CharField(max_length=255)
    country = CountryField(default="CH")
    description = LocalizedCharField(blank=True, null=True, required=False)
    default = UniqueBooleanField(default=False, together=["identity"])

    class Meta:
        ordering = ("-default",)
