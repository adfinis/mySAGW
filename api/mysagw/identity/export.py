from collections import OrderedDict


def _str_or_empty(value):
    mapping = {True: 1, False: 0, None: ""}
    return mapping.get(value, str(value))


def _fetch_related_strings(manager, field):
    return [_str_or_empty(getattr(entry, field)) for entry in manager.iterator()]


class IdentityExport:
    FIELDS = {
        "first_name": lambda i, a: i.first_name,
        "last_name": lambda i, a: i.last_name,
        "localized_salutation": lambda i, a: i.localized_salutation,
        "localized_title": lambda i, a: i.localized_title,
        "language": lambda i, a: i.language,
        "is_organisation": lambda i, a: i.is_organisation,
        "organisation_name": lambda i, a: i.organisation_name,
        "email": lambda i, a: i.email,
        "additional_emails": lambda i, a: "\n".join(
            _fetch_related_strings(i.additional_emails, "email"),
        ),
        "phone_numbers": lambda i, a: "\n".join(
            _fetch_related_strings(i.phone_numbers, "phone"),
        ),
        "address_addition_1": lambda i, a: "" if not a else a.address_addition_1,
        "address_addition_2": lambda i, a: "" if not a else a.address_addition_2,
        "address_addition_3": lambda i, a: "" if not a else a.address_addition_3,
        "street_and_number": lambda i, a: "" if not a else a.street_and_number,
        "po_box": lambda i, a: "" if not a else a.po_box,
        "postcode": lambda i, a: "" if not a else a.postcode,
        "town": lambda i, a: "" if not a else a.town,
        "country": lambda i, a: "" if not a else a.country.name,
    }

    def fetch_identity_data(self, identity, fields):
        address = None
        if identity.addresses.exists():
            address = identity.addresses.get(default=True)

        data = OrderedDict()

        for field_key, field_getter in fields.items():
            data[field_key] = _str_or_empty(field_getter(identity, address))

        return data

    def export(self, qs, include_fields=None, ignore_empty=False):
        include_fields = include_fields if include_fields else self.FIELDS.keys()
        fields = {name: self.FIELDS[name] for name in include_fields}

        records = []
        for identity in qs.iterator(2000):
            identity_data = self.fetch_identity_data(identity, fields)

            if not ignore_empty or any(identity_data.values()):
                records.append(identity_data)

        return records


class MembershipExport:
    FIELDS = {
        "first_name": lambda m, a: m.identity.first_name,
        "last_name": lambda m, a: m.identity.last_name,
        "localized_salutation": lambda m, a: m.identity.localized_salutation,
        "localized_title": lambda m, a: m.identity.localized_title,
        "language": lambda m, a: m.identity.language,
        "email": lambda m, a: m.identity.email,
        "address_addition_1": lambda m, a: "" if not a else a.address_addition_1,
        "address_addition_2": lambda m, a: "" if not a else a.address_addition_2,
        "address_addition_3": lambda m, a: "" if not a else a.address_addition_3,
        "street_and_number": lambda m, a: "" if not a else a.street_and_number,
        "po_box": lambda m, a: "" if not a else a.po_box,
        "postcode": lambda m, a: "" if not a else a.postcode,
        "town": lambda m, a: "" if not a else a.town,
        "country": lambda m, a: "" if not a else a.country.name,
        "organisation": lambda m, a: m.organisation.organisation_name,
        "role": lambda m, a: "" if not m.role else m.role.title[m.identity.language],
        "inactive": lambda m, a: m.inactive,
        "tenure": lambda m, a: ""
        if not m.time_slot
        else f"{m.time_slot.lower if m.time_slot.lower else ''} - {m.time_slot.upper if m.time_slot.upper else ''}",
        "next_election": lambda m, a: m.next_election,
    }

    def fetch_membership_data(self, membership, fields):
        address = None
        if membership.identity.addresses.exists():
            address = membership.identity.addresses.get(default=True)

        data = OrderedDict()

        for field_key, field_getter in fields.items():
            data[field_key] = _str_or_empty(field_getter(membership, address))

        return data

    def export(self, qs, include_fields=None, ignore_empty=False):
        include_fields = include_fields if include_fields else self.FIELDS.keys()
        fields = {name: self.FIELDS[name] for name in include_fields}

        records = []
        for membership in qs.iterator(2000):
            membership_data = self.fetch_membership_data(membership, fields)

            if not ignore_empty or any(membership_data.values()):
                records.append(membership_data)

        return records
