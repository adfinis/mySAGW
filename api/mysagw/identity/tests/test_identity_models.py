import pytest
from django.conf import settings

from mysagw.identity.models import Identity


@pytest.mark.parametrize(
    "identity__salutation", Identity.SALUTATION_LOCALIZED_MAP.keys()
)
@pytest.mark.parametrize("identity__title", Identity.TITLE_LOCALIZED_MAP.keys())
@pytest.mark.parametrize("identity__first_name", ["Winston", None])
@pytest.mark.parametrize("identity__last_name", ["Smith", None])
@pytest.mark.parametrize(
    "identity__language", settings.LOCALIZED_FIELDS_FALLBACKS.keys()
)
def test_fullname(db, identity, snapshot):
    snapshot.assert_match(identity.full_name)


@pytest.mark.parametrize(
    "address__address_addition_1, address__address_addition_2, address__address_addition_3",
    [
        (None, None, None),
        ("Some", "Nice", "Addition"),
        (None, "Nice", "Addition"),
        ("Some", None, "Addition"),
    ],
)
@pytest.mark.parametrize("lang", settings.LOCALIZED_FIELDS_FALLBACKS.keys())
@pytest.mark.parametrize("address__po_box", ["23", None])
def test_address_block(db, snapshot, address, lang):
    address.identity.language = lang
    address.identity.save()
    snapshot.assert_match(address.identity.address_block)


def test_address_block_missing_address(db, snapshot, identity):
    snapshot.assert_match(identity.address_block)
