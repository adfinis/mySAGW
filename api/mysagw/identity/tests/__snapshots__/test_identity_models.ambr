# serializer version: 1
# name: test_address_block[23-de-None-Nice-Addition]
  '''
  David Moreno
  Nice
  Addition
  060 Guerra Island
  23
  62422 Port Michaela
  Äthiopien
  '''
# ---
# name: test_address_block[23-de-None-None-None]
  '''
  David Moreno
  060 Guerra Island
  23
  62422 Port Michaela
  Äthiopien
  '''
# ---
# name: test_address_block[23-de-Some-Nice-Addition]
  '''
  David Moreno
  Some
  Nice
  Addition
  060 Guerra Island
  23
  62422 Port Michaela
  Äthiopien
  '''
# ---
# name: test_address_block[23-de-Some-None-Addition]
  '''
  David Moreno
  Some
  Addition
  060 Guerra Island
  23
  62422 Port Michaela
  Äthiopien
  '''
# ---
# name: test_address_block[23-en-None-Nice-Addition]
  '''
  David Moreno
  Nice
  Addition
  060 Guerra Island
  23
  62422 Port Michaela
  Ethiopia
  '''
# ---
# name: test_address_block[23-en-None-None-None]
  '''
  David Moreno
  060 Guerra Island
  23
  62422 Port Michaela
  Ethiopia
  '''
# ---
# name: test_address_block[23-en-Some-Nice-Addition]
  '''
  David Moreno
  Some
  Nice
  Addition
  060 Guerra Island
  23
  62422 Port Michaela
  Ethiopia
  '''
# ---
# name: test_address_block[23-en-Some-None-Addition]
  '''
  David Moreno
  Some
  Addition
  060 Guerra Island
  23
  62422 Port Michaela
  Ethiopia
  '''
# ---
# name: test_address_block[23-fr-None-Nice-Addition]
  '''
  David Moreno
  Nice
  Addition
  060 Guerra Island
  23
  62422 Port Michaela
  Éthiopie
  '''
# ---
# name: test_address_block[23-fr-None-None-None]
  '''
  David Moreno
  060 Guerra Island
  23
  62422 Port Michaela
  Éthiopie
  '''
# ---
# name: test_address_block[23-fr-Some-Nice-Addition]
  '''
  David Moreno
  Some
  Nice
  Addition
  060 Guerra Island
  23
  62422 Port Michaela
  Éthiopie
  '''
# ---
# name: test_address_block[23-fr-Some-None-Addition]
  '''
  David Moreno
  Some
  Addition
  060 Guerra Island
  23
  62422 Port Michaela
  Éthiopie
  '''
# ---
# name: test_address_block[None-de-None-Nice-Addition]
  '''
  David Moreno
  Nice
  Addition
  060 Guerra Island
  62422 Port Michaela
  Äthiopien
  '''
# ---
# name: test_address_block[None-de-None-None-None]
  '''
  David Moreno
  060 Guerra Island
  62422 Port Michaela
  Äthiopien
  '''
# ---
# name: test_address_block[None-de-Some-Nice-Addition]
  '''
  David Moreno
  Some
  Nice
  Addition
  060 Guerra Island
  62422 Port Michaela
  Äthiopien
  '''
# ---
# name: test_address_block[None-de-Some-None-Addition]
  '''
  David Moreno
  Some
  Addition
  060 Guerra Island
  62422 Port Michaela
  Äthiopien
  '''
# ---
# name: test_address_block[None-en-None-Nice-Addition]
  '''
  David Moreno
  Nice
  Addition
  060 Guerra Island
  62422 Port Michaela
  Ethiopia
  '''
# ---
# name: test_address_block[None-en-None-None-None]
  '''
  David Moreno
  060 Guerra Island
  62422 Port Michaela
  Ethiopia
  '''
# ---
# name: test_address_block[None-en-Some-Nice-Addition]
  '''
  David Moreno
  Some
  Nice
  Addition
  060 Guerra Island
  62422 Port Michaela
  Ethiopia
  '''
# ---
# name: test_address_block[None-en-Some-None-Addition]
  '''
  David Moreno
  Some
  Addition
  060 Guerra Island
  62422 Port Michaela
  Ethiopia
  '''
# ---
# name: test_address_block[None-fr-None-Nice-Addition]
  '''
  David Moreno
  Nice
  Addition
  060 Guerra Island
  62422 Port Michaela
  Éthiopie
  '''
# ---
# name: test_address_block[None-fr-None-None-None]
  '''
  David Moreno
  060 Guerra Island
  62422 Port Michaela
  Éthiopie
  '''
# ---
# name: test_address_block[None-fr-Some-Nice-Addition]
  '''
  David Moreno
  Some
  Nice
  Addition
  060 Guerra Island
  62422 Port Michaela
  Éthiopie
  '''
# ---
# name: test_address_block[None-fr-Some-None-Addition]
  '''
  David Moreno
  Some
  Addition
  060 Guerra Island
  62422 Port Michaela
  Éthiopie
  '''
# ---
# name: test_address_block_missing_address
  'David Moreno'
# ---
# name: test_fullname[de-None-None-dr-female]
  'Frau Dr.'
# ---
# name: test_fullname[de-None-None-dr-male]
  'Herr Dr.'
# ---
# name: test_fullname[de-None-None-dr-neutral]
  'Dr.'
# ---
# name: test_fullname[de-None-None-none-female]
  'Frau'
# ---
# name: test_fullname[de-None-None-none-male]
  'Herr'
# ---
# name: test_fullname[de-None-None-none-neutral]
  ''
# ---
# name: test_fullname[de-None-None-pd-dr-female]
  'Frau PD Dr.'
# ---
# name: test_fullname[de-None-None-pd-dr-male]
  'Herr PD Dr.'
# ---
# name: test_fullname[de-None-None-pd-dr-neutral]
  'PD Dr.'
# ---
# name: test_fullname[de-None-None-prof-dr-female]
  'Frau Prof. Dr.'
# ---
# name: test_fullname[de-None-None-prof-dr-male]
  'Herr Prof. Dr.'
# ---
# name: test_fullname[de-None-None-prof-dr-neutral]
  'Prof. Dr.'
# ---
# name: test_fullname[de-None-None-prof-female]
  'Frau Prof.'
# ---
# name: test_fullname[de-None-None-prof-male]
  'Herr Prof.'
# ---
# name: test_fullname[de-None-None-prof-neutral]
  'Prof.'
# ---
# name: test_fullname[de-None-Winston-dr-female]
  'Frau Dr. Winston'
# ---
# name: test_fullname[de-None-Winston-dr-male]
  'Herr Dr. Winston'
# ---
# name: test_fullname[de-None-Winston-dr-neutral]
  'Dr. Winston'
# ---
# name: test_fullname[de-None-Winston-none-female]
  'Frau Winston'
# ---
# name: test_fullname[de-None-Winston-none-male]
  'Herr Winston'
# ---
# name: test_fullname[de-None-Winston-none-neutral]
  'Winston'
# ---
# name: test_fullname[de-None-Winston-pd-dr-female]
  'Frau PD Dr. Winston'
# ---
# name: test_fullname[de-None-Winston-pd-dr-male]
  'Herr PD Dr. Winston'
# ---
# name: test_fullname[de-None-Winston-pd-dr-neutral]
  'PD Dr. Winston'
# ---
# name: test_fullname[de-None-Winston-prof-dr-female]
  'Frau Prof. Dr. Winston'
# ---
# name: test_fullname[de-None-Winston-prof-dr-male]
  'Herr Prof. Dr. Winston'
# ---
# name: test_fullname[de-None-Winston-prof-dr-neutral]
  'Prof. Dr. Winston'
# ---
# name: test_fullname[de-None-Winston-prof-female]
  'Frau Prof. Winston'
# ---
# name: test_fullname[de-None-Winston-prof-male]
  'Herr Prof. Winston'
# ---
# name: test_fullname[de-None-Winston-prof-neutral]
  'Prof. Winston'
# ---
# name: test_fullname[de-Smith-None-dr-female]
  'Frau Dr. Smith'
# ---
# name: test_fullname[de-Smith-None-dr-male]
  'Herr Dr. Smith'
# ---
# name: test_fullname[de-Smith-None-dr-neutral]
  'Dr. Smith'
# ---
# name: test_fullname[de-Smith-None-none-female]
  'Frau Smith'
# ---
# name: test_fullname[de-Smith-None-none-male]
  'Herr Smith'
# ---
# name: test_fullname[de-Smith-None-none-neutral]
  'Smith'
# ---
# name: test_fullname[de-Smith-None-pd-dr-female]
  'Frau PD Dr. Smith'
# ---
# name: test_fullname[de-Smith-None-pd-dr-male]
  'Herr PD Dr. Smith'
# ---
# name: test_fullname[de-Smith-None-pd-dr-neutral]
  'PD Dr. Smith'
# ---
# name: test_fullname[de-Smith-None-prof-dr-female]
  'Frau Prof. Dr. Smith'
# ---
# name: test_fullname[de-Smith-None-prof-dr-male]
  'Herr Prof. Dr. Smith'
# ---
# name: test_fullname[de-Smith-None-prof-dr-neutral]
  'Prof. Dr. Smith'
# ---
# name: test_fullname[de-Smith-None-prof-female]
  'Frau Prof. Smith'
# ---
# name: test_fullname[de-Smith-None-prof-male]
  'Herr Prof. Smith'
# ---
# name: test_fullname[de-Smith-None-prof-neutral]
  'Prof. Smith'
# ---
# name: test_fullname[de-Smith-Winston-dr-female]
  'Frau Dr. Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-dr-male]
  'Herr Dr. Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-dr-neutral]
  'Dr. Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-none-female]
  'Frau Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-none-male]
  'Herr Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-none-neutral]
  'Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-pd-dr-female]
  'Frau PD Dr. Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-pd-dr-male]
  'Herr PD Dr. Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-pd-dr-neutral]
  'PD Dr. Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-prof-dr-female]
  'Frau Prof. Dr. Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-prof-dr-male]
  'Herr Prof. Dr. Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-prof-dr-neutral]
  'Prof. Dr. Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-prof-female]
  'Frau Prof. Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-prof-male]
  'Herr Prof. Winston Smith'
# ---
# name: test_fullname[de-Smith-Winston-prof-neutral]
  'Prof. Winston Smith'
# ---
# name: test_fullname[en-None-None-dr-female]
  'Mrs. Dr.'
# ---
# name: test_fullname[en-None-None-dr-male]
  'Mr. Dr.'
# ---
# name: test_fullname[en-None-None-dr-neutral]
  'Dr.'
# ---
# name: test_fullname[en-None-None-none-female]
  'Mrs.'
# ---
# name: test_fullname[en-None-None-none-male]
  'Mr.'
# ---
# name: test_fullname[en-None-None-none-neutral]
  ''
# ---
# name: test_fullname[en-None-None-pd-dr-female]
  'Mrs. PD Dr.'
# ---
# name: test_fullname[en-None-None-pd-dr-male]
  'Mr. PD Dr.'
# ---
# name: test_fullname[en-None-None-pd-dr-neutral]
  'PD Dr.'
# ---
# name: test_fullname[en-None-None-prof-dr-female]
  'Mrs. Prof. Dr.'
# ---
# name: test_fullname[en-None-None-prof-dr-male]
  'Mr. Prof. Dr.'
# ---
# name: test_fullname[en-None-None-prof-dr-neutral]
  'Prof. Dr.'
# ---
# name: test_fullname[en-None-None-prof-female]
  'Mrs. Prof.'
# ---
# name: test_fullname[en-None-None-prof-male]
  'Mr. Prof.'
# ---
# name: test_fullname[en-None-None-prof-neutral]
  'Prof.'
# ---
# name: test_fullname[en-None-Winston-dr-female]
  'Mrs. Dr. Winston'
# ---
# name: test_fullname[en-None-Winston-dr-male]
  'Mr. Dr. Winston'
# ---
# name: test_fullname[en-None-Winston-dr-neutral]
  'Dr. Winston'
# ---
# name: test_fullname[en-None-Winston-none-female]
  'Mrs. Winston'
# ---
# name: test_fullname[en-None-Winston-none-male]
  'Mr. Winston'
# ---
# name: test_fullname[en-None-Winston-none-neutral]
  'Winston'
# ---
# name: test_fullname[en-None-Winston-pd-dr-female]
  'Mrs. PD Dr. Winston'
# ---
# name: test_fullname[en-None-Winston-pd-dr-male]
  'Mr. PD Dr. Winston'
# ---
# name: test_fullname[en-None-Winston-pd-dr-neutral]
  'PD Dr. Winston'
# ---
# name: test_fullname[en-None-Winston-prof-dr-female]
  'Mrs. Prof. Dr. Winston'
# ---
# name: test_fullname[en-None-Winston-prof-dr-male]
  'Mr. Prof. Dr. Winston'
# ---
# name: test_fullname[en-None-Winston-prof-dr-neutral]
  'Prof. Dr. Winston'
# ---
# name: test_fullname[en-None-Winston-prof-female]
  'Mrs. Prof. Winston'
# ---
# name: test_fullname[en-None-Winston-prof-male]
  'Mr. Prof. Winston'
# ---
# name: test_fullname[en-None-Winston-prof-neutral]
  'Prof. Winston'
# ---
# name: test_fullname[en-Smith-None-dr-female]
  'Mrs. Dr. Smith'
# ---
# name: test_fullname[en-Smith-None-dr-male]
  'Mr. Dr. Smith'
# ---
# name: test_fullname[en-Smith-None-dr-neutral]
  'Dr. Smith'
# ---
# name: test_fullname[en-Smith-None-none-female]
  'Mrs. Smith'
# ---
# name: test_fullname[en-Smith-None-none-male]
  'Mr. Smith'
# ---
# name: test_fullname[en-Smith-None-none-neutral]
  'Smith'
# ---
# name: test_fullname[en-Smith-None-pd-dr-female]
  'Mrs. PD Dr. Smith'
# ---
# name: test_fullname[en-Smith-None-pd-dr-male]
  'Mr. PD Dr. Smith'
# ---
# name: test_fullname[en-Smith-None-pd-dr-neutral]
  'PD Dr. Smith'
# ---
# name: test_fullname[en-Smith-None-prof-dr-female]
  'Mrs. Prof. Dr. Smith'
# ---
# name: test_fullname[en-Smith-None-prof-dr-male]
  'Mr. Prof. Dr. Smith'
# ---
# name: test_fullname[en-Smith-None-prof-dr-neutral]
  'Prof. Dr. Smith'
# ---
# name: test_fullname[en-Smith-None-prof-female]
  'Mrs. Prof. Smith'
# ---
# name: test_fullname[en-Smith-None-prof-male]
  'Mr. Prof. Smith'
# ---
# name: test_fullname[en-Smith-None-prof-neutral]
  'Prof. Smith'
# ---
# name: test_fullname[en-Smith-Winston-dr-female]
  'Mrs. Dr. Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-dr-male]
  'Mr. Dr. Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-dr-neutral]
  'Dr. Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-none-female]
  'Mrs. Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-none-male]
  'Mr. Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-none-neutral]
  'Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-pd-dr-female]
  'Mrs. PD Dr. Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-pd-dr-male]
  'Mr. PD Dr. Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-pd-dr-neutral]
  'PD Dr. Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-prof-dr-female]
  'Mrs. Prof. Dr. Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-prof-dr-male]
  'Mr. Prof. Dr. Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-prof-dr-neutral]
  'Prof. Dr. Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-prof-female]
  'Mrs. Prof. Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-prof-male]
  'Mr. Prof. Winston Smith'
# ---
# name: test_fullname[en-Smith-Winston-prof-neutral]
  'Prof. Winston Smith'
# ---
# name: test_fullname[fr-None-None-dr-female]
  'Madame Dr'
# ---
# name: test_fullname[fr-None-None-dr-male]
  'Monsieur Dr'
# ---
# name: test_fullname[fr-None-None-dr-neutral]
  'Dr'
# ---
# name: test_fullname[fr-None-None-none-female]
  'Madame'
# ---
# name: test_fullname[fr-None-None-none-male]
  'Monsieur'
# ---
# name: test_fullname[fr-None-None-none-neutral]
  ''
# ---
# name: test_fullname[fr-None-None-pd-dr-female]
  'Madame PD Dr'
# ---
# name: test_fullname[fr-None-None-pd-dr-male]
  'Monsieur PD Dr'
# ---
# name: test_fullname[fr-None-None-pd-dr-neutral]
  'PD Dr'
# ---
# name: test_fullname[fr-None-None-prof-dr-female]
  'Madame Prof. Dr'
# ---
# name: test_fullname[fr-None-None-prof-dr-male]
  'Monsieur Prof. Dr'
# ---
# name: test_fullname[fr-None-None-prof-dr-neutral]
  'Prof. Dr'
# ---
# name: test_fullname[fr-None-None-prof-female]
  'Madame Prof.'
# ---
# name: test_fullname[fr-None-None-prof-male]
  'Monsieur Prof.'
# ---
# name: test_fullname[fr-None-None-prof-neutral]
  'Prof.'
# ---
# name: test_fullname[fr-None-Winston-dr-female]
  'Madame Dr Winston'
# ---
# name: test_fullname[fr-None-Winston-dr-male]
  'Monsieur Dr Winston'
# ---
# name: test_fullname[fr-None-Winston-dr-neutral]
  'Dr Winston'
# ---
# name: test_fullname[fr-None-Winston-none-female]
  'Madame Winston'
# ---
# name: test_fullname[fr-None-Winston-none-male]
  'Monsieur Winston'
# ---
# name: test_fullname[fr-None-Winston-none-neutral]
  'Winston'
# ---
# name: test_fullname[fr-None-Winston-pd-dr-female]
  'Madame PD Dr Winston'
# ---
# name: test_fullname[fr-None-Winston-pd-dr-male]
  'Monsieur PD Dr Winston'
# ---
# name: test_fullname[fr-None-Winston-pd-dr-neutral]
  'PD Dr Winston'
# ---
# name: test_fullname[fr-None-Winston-prof-dr-female]
  'Madame Prof. Dr Winston'
# ---
# name: test_fullname[fr-None-Winston-prof-dr-male]
  'Monsieur Prof. Dr Winston'
# ---
# name: test_fullname[fr-None-Winston-prof-dr-neutral]
  'Prof. Dr Winston'
# ---
# name: test_fullname[fr-None-Winston-prof-female]
  'Madame Prof. Winston'
# ---
# name: test_fullname[fr-None-Winston-prof-male]
  'Monsieur Prof. Winston'
# ---
# name: test_fullname[fr-None-Winston-prof-neutral]
  'Prof. Winston'
# ---
# name: test_fullname[fr-Smith-None-dr-female]
  'Madame Dr Smith'
# ---
# name: test_fullname[fr-Smith-None-dr-male]
  'Monsieur Dr Smith'
# ---
# name: test_fullname[fr-Smith-None-dr-neutral]
  'Dr Smith'
# ---
# name: test_fullname[fr-Smith-None-none-female]
  'Madame Smith'
# ---
# name: test_fullname[fr-Smith-None-none-male]
  'Monsieur Smith'
# ---
# name: test_fullname[fr-Smith-None-none-neutral]
  'Smith'
# ---
# name: test_fullname[fr-Smith-None-pd-dr-female]
  'Madame PD Dr Smith'
# ---
# name: test_fullname[fr-Smith-None-pd-dr-male]
  'Monsieur PD Dr Smith'
# ---
# name: test_fullname[fr-Smith-None-pd-dr-neutral]
  'PD Dr Smith'
# ---
# name: test_fullname[fr-Smith-None-prof-dr-female]
  'Madame Prof. Dr Smith'
# ---
# name: test_fullname[fr-Smith-None-prof-dr-male]
  'Monsieur Prof. Dr Smith'
# ---
# name: test_fullname[fr-Smith-None-prof-dr-neutral]
  'Prof. Dr Smith'
# ---
# name: test_fullname[fr-Smith-None-prof-female]
  'Madame Prof. Smith'
# ---
# name: test_fullname[fr-Smith-None-prof-male]
  'Monsieur Prof. Smith'
# ---
# name: test_fullname[fr-Smith-None-prof-neutral]
  'Prof. Smith'
# ---
# name: test_fullname[fr-Smith-Winston-dr-female]
  'Madame Dr Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-dr-male]
  'Monsieur Dr Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-dr-neutral]
  'Dr Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-none-female]
  'Madame Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-none-male]
  'Monsieur Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-none-neutral]
  'Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-pd-dr-female]
  'Madame PD Dr Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-pd-dr-male]
  'Monsieur PD Dr Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-pd-dr-neutral]
  'PD Dr Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-prof-dr-female]
  'Madame Prof. Dr Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-prof-dr-male]
  'Monsieur Prof. Dr Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-prof-dr-neutral]
  'Prof. Dr Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-prof-female]
  'Madame Prof. Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-prof-male]
  'Monsieur Prof. Winston Smith'
# ---
# name: test_fullname[fr-Smith-Winston-prof-neutral]
  'Prof. Winston Smith'
# ---
